pkg/swap/swap.sh

```diff
@@ -0,0 +1,122 @@
+#!/bin/bash
+
+set -e
+
+while [[ $# -gt 1 ]]; do
+	key="$1"
+
+	case $key in
+		--debug)
+			set -x
+			;;
+    --path)
+	    path="$2"
+	    shift # past argument
+	    ;;
+    --size)
+	    size="$2"
+	    shift # past argument
+	    ;;
+    --encrypt)
+	    ENCRYPT=true
+	    ;;
+		--condition)
+	    CONDITIONS="$CONDITIONS $2"
+			shift
+	    ;;
+    *)
+      echo "Unknown option passed to swapmaker: $key"      # unknown option
+			exit 1
+	    ;;
+	esac
+	shift # past argument or value
+done
+
+function minsize_to_bytes {
+	local origsize=$1
+	local ret
+	case $origsize in
+		*G)
+			ret=$(( ${origsize%%G} * 1024 * 1024 * 1024 ))
+			;;
+		*M)
+			ret=$(( ${origsize%%M} * 1024 * 1024 ))
+			;;
+		*K)
+			ret=$(( ${origsize%%K} * 1024 ))
+			;;
+		*)
+			ret=$origsize
+			;;
+	esac
+	echo $ret
+}
+
+## make sure path is valid
+if ! [[ $path =~ ^/var/.+$ ]]; then
+	echo "--file <path> option must be under /var"
+	exit 1
+fi
+
+## check each of our conditions
+for cond in $CONDITIONS; do
+	# split the condition parts
+	IFS=: read -a parts <<< $cond
+	case ${parts[0]} in
+		part)
+			partition=${parts[1]}
+			required=${parts[2]}
+			# check that the path exists as its own mount point
+			set +e
+			grep -qs $partition /proc/mounts
+			is_mnt=$?
+			set -e
+			if [[ $is_mnt != 0 ]]; then
+				[[ "$required" == "true" ]] && exit 1
+				exit 0
+			fi
+			;;
+		partsize)
+			partition=${parts[1]}
+			minsize=${parts[2]}
+			required=${parts[3]}
+			# check that the partition on which it exists has sufficient size
+			partsize=$(df -k $partition | tail -1 | awk '{print $2}')
+			partsize=$(( $partsize * 1024 ))
+			# convert minsize to bytes
+			minsize=$(minsize_to_bytes $minsize)
+			if [[ $partsize -lt $minsize ]]; then
+				[[ "$required" == "true" ]] && exit 1
+				exit 0
+			fi
+			;;
+		*)
+			echo "Unknown condition: $cond"
+			exit 1
+			;;
+	esac
+done
+## if a condition failed:
+### Required? exit 1
+### Else? exit 0
+
+
+## Allocate the file
+fallocate -l $size $path
```

Comment:
I think the advantage is you get consecutive allocations if you do it up front with `dd` (which is why it is slow). On spinning rust that was important. On SSD it probably helps a bit too as real write units are 128k or so too. So probably still worth paying the cost upfront.

On Docker for mac we only create the file if it doesnt already exist so you only pay the cost on first boot.